### PR TITLE
Split core runtime from kotlin runtime

### PR DESCRIPTION
**Pull request by ericvergnaud**
This PR:
 - creates a new module 'core', that will be later compiled to Wasm (and maybe Jvm), moving code from the kotlin runtime
 - keeps with the kotlin runtime any stuff that's i/o related
 - normalizes source directories

This is phase 1. Upcoming phases will:
 - have the Java runtime use the Core runtime
 - switch to Gradle
 - create the API exposed by the Wasm

### CODE REVIEW

File: runtime/Java/src/org/antlr/v5/runtime/ANTLRFileStream.java
Comment (KvanTTT) [inline runtime/Java/src/org/antlr/v5/runtime/ANTLRFileStream.java:1]: This change looks unrelated. How does it block splitting kotlin runtime?

File: runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/tree/xpath/XPathTokenAnywhereElement.kt [hunk 1]
```diff
@@ -3,10 +3,10 @@
  * Use of this file is governed by the BSD 3-clause license that
  * can be found in the LICENSE.txt file in the project root.
  */
-package org.antlr.v5.runtime.kotlin.tree.xpath
+package org.antlr.v5.runtime.kotlin._unused.tree.xpath
```
Comment (KvanTTT) [inline runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/tree/xpath/XPathTokenAnywhereElement.kt:6]: What's the meaning for `_unused` part?

File: runtime/Java/pom.xml [hunk 2]
```diff
@@ -23,10 +23,25 @@
...
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-engine</artifactId>
+            <version>${junit.version}</version>
+            <scope>test</scope>
+        </dependency>
+
+    </dependencies>
```
Comment (KvanTTT) [inline runtime/Java/pom.xml:43]: These dependencies are incorrect for `runtime/Java` module. It doesn't test anything.

File: runtime/Java/src/test/java/org/antlr/v5/runtime/TestCodePointCharStream.java [hunk 3]
```diff
@@ -7,173 +7,173 @@
...
-import static org.junit.jupiter.api.Assertions.*;
 
 public class TestCodePointCharStream {
 	@Test
 	public void emptyBytesHasSize0() {
 		CodePointCharStream s = CharStreams.fromString("");
-		assertEquals(0, s.size());
-		assertEquals(0, s.index());
-		assertEquals("", s.toString());
+		Assertions.assertEquals(0, s.size());
```
Comment (KvanTTT) [inline runtime/Java/src/test/java/org/antlr/v5/runtime/TestCodePointCharStream.java]: Unrelated changes.

File: tool/resources/org/antlr/v5/tool/templates/codegen/Kotlin/Kotlin.stg [hunk 4]
```diff
@@ -226,6 +229,11 @@ Parser(parser, funcs, atn, sempredFuncs, superClass) ::= <<
 Parser_(parser, funcs, atn, sempredFuncs, superClass) ::= <<
 @Suppress("UNUSED_VARIABLE", "ClassName", "FunctionName", "LocalVariableName", "ConstPropertyName", "ConvertSecondaryConstructorToPrimary")
 public open class <parser.name>(input: TokenStream) : <superClass; null="Parser">(input) {
+
+	init {
+		addErrorListener(ConsoleErrorListener())
```
Comment (KvanTTT) [inline tool/resources/org/antlr/v5/tool/templates/codegen/Kotlin/Kotlin.stg:234]: Why was this line added?

File: runtime/Core/src/main/kotlin/org/antlr/v5/runtime/core/IPrintStream.kt [hunk 5]
```diff
@@ -0,0 +1,8 @@
+package org.antlr.v5.runtime.core
+
+public interface IPrintStream {
+
+    public fun print(value: String)
+    public fun printLine(value: String?)
```
Comment (KvanTTT) [inline runtime/Core/src/main/kotlin/org/antlr/v5/runtime/core/IPrintStream.kt:6]: I suggest use conventional name `println` instead of `printLine`.

File: runtime-testsuite/test/org/antlr/v5/test/runtime/java/api/TestVisitors.java [hunk 6]
```diff
@@ -30,7 +26,7 @@ public class TestVisitors {
 	@Test
 	public void testVisitTerminalNode() {
 		String input = "A";
-		VisitorBasicLexer lexer = new VisitorBasicLexer(new ANTLRInputStream(input));
+		VisitorBasicLexer lexer = new VisitorBasicLexer(CharStreams.fromString(input));
```
Comment (KvanTTT) [inline runtime-testsuite/test/org/antlr/v5/test/runtime/java/api/TestVisitors.java:29]: It looks like unrelated changes.

**Review by KvanTTT — CHANGES_REQUESTED** [review]
> creates a new module 'core', that will be later compiled to Wasm (and maybe Jvm), moving code from the kotlin runtime

It should have another structure: `core` (or maybe even `common` or `kotlin-common`) should be a common module with some `expect` declarations. `wasm`, `jvm`, `native` should be separated modules with respective `actual` declarations (that's why we still have some files in [antlr-kotlin-runtime-remnants](https://github.com/antlr/antlr5/tree/dev/antlr-kotlin-runtime-remnants) that should be ported to respective modules).

> have the Java runtime use the Core runtime

Current structure doesn't prevent of doing that. I've already tried it but encountered a lot of problems regarding naming issues: Kotlin relies on properties but Java doesn't have them. Also, I don't think this task has high-priority since currently we can't port it to Wasm because it has some outdated dependencies that should be eliminated at first. It requires a lot of effort.

> switch to Gradle

I suggest starting from migration to Gradle because Maven doesn't support KMP features. Without KMP features we can't test both JVM and Wasm.

Also, all checks are failed and a lot number of changes probably will force me to resolve conflicts in my pending MRs again: https://github.com/antlr/antlr5/pull/30, https://github.com/antlr/antlr5/pull/31. Please let me merge them at first.

**Comment by KvanTTT** [comment]
Please leave this splitting to me, If you don't object. Probably I have more experience regarding Kotlin KMP.

File: runtime/Java/pom.xml (hunk 2 above)
Comment (ericvergnaud) [reply runtime/Java/pom.xml:43]: It does now

File: runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/tree/xpath/XPathTokenAnywhereElement.kt (hunk 1 above)
Comment (ericvergnaud) [reply runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/tree/xpath/XPathTokenAnywhereElement.kt:6]: stuff that's in the code base but is not used within. This is notably the xpath stuff that we want to externalize.

File: runtime/Java/src/org/antlr/v5/runtime/ANTLRFileStream.java
Comment (ericvergnaud) [reply runtime/Java/src/org/antlr/v5/runtime/ANTLRFileStream.java:1]: It is unrelated, it was marked as deprecated a long time ago, so it's housekeeping

File: tool/resources/org/antlr/v5/tool/templates/codegen/Kotlin/Kotlin.stg (hunk 4 above)
Comment (ericvergnaud) [reply tool/resources/org/antlr/v5/tool/templates/codegen/Kotlin/Kotlin.stg:234]: because originally the default error listener is the ConsoleErrorListener, which cannot be part of core because it requires IO. This ensures a backwards compatible behavior

File: runtime/Core/src/main/kotlin/org/antlr/v5/runtime/core/IPrintStream.kt (hunk 5 above)
Comment (ericvergnaud) [reply runtime/Core/src/main/kotlin/org/antlr/v5/runtime/core/IPrintStream.kt:6]: well println is conventional in the java world only. This will be part of the API so it needs to be less sticky

File: runtime-testsuite/test/org/antlr/v5/test/runtime/java/api/TestVisitors.java (hunk 6 above)
Comment (ericvergnaud) [reply runtime-testsuite/test/org/antlr/v5/test/runtime/java/api/TestVisitors.java:29]: It relates to housekeeping, see above.
Since I had to revisit many files, I took the opportunity.

**Comment by ericvergnaud** [comment]
> Please leave this splitting to me, If you don't object. Probably I have more experience regarding Kotlin KMP.

You definitely have more experience with Kotlin and KMP.
But this isn't about Kotlin or KMP yet. It's about our target architecture.
That said you're welcome to help with the failing build, it works locally, o not sure why it fails in the CI...

File: runtime/Java/pom.xml (hunk 2 above)
Comment (KvanTTT) [reply runtime/Java/pom.xml:43]: But currently runtime (`runtime/Java`) and tests (`tool-testsuite`, `runtime-testsuite`) modules are currently separated, I don't see a reason why tests should be placed in all these modules. It looks inconsistent. Please move all tests (I guess `TestCodePointCharStream`) to `tool-testsuite` or `runtime-testsuite`.

**Comment by ericvergnaud** [comment]
> > creates a new module 'core', that will be later compiled to Wasm (and maybe Jvm), moving code from the kotlin runtime
> 
> It should have another structure: `core` (or maybe even `common` or `kotlin-common`) should be a common module with some `expect` declarations. `wasm`, `jvm`, `native` should be separated modules with respective `actual` declarations (that's why we still have some files in [antlr-kotlin-runtime-remnants](https://github.com/antlr/antlr5/tree/dev/antlr-kotlin-runtime-remnants) that should be ported to respective modules).
> 
> > have the Java runtime use the Core runtime
> 
> Current structure doesn't prevent of doing that. I've already tried it but encountered a lot of problems regarding naming issues: Kotlin relies on properties but Java doesn't have them. Also, I don't think this task has high-priority since currently we can't port it to Wasm because it has some outdated dependencies that should be eliminated at first. It requires a lot of effort.
> 
> > switch to Gradle
> 
> I suggest starting from migration to Gradle because Maven doesn't support KMP features. Without KMP features we can't test both JVM and Wasm.
> 
> Also, all checks are failed and a lot number of changes probably will force me to resolve conflicts in my pending MRs again: #30, #31. Please let me merge them at first.

I'm happy to merge #30 and #31 before this but they don't show as ready ?

Doing this before KMP is much more effective than the other way around (it's ok to think differently about this). We'll enable KMP (and deal with all the above remarks) as soon as this re-architecting is complete.

File: runtime-testsuite/test/org/antlr/v5/test/runtime/java/api/TestVisitors.java (hunk 6 above)
Comment (KvanTTT) [reply runtime-testsuite/test/org/antlr/v5/test/runtime/java/api/TestVisitors.java:29]: Please try to place different changes to different PR next time. It confuses reviewing process especially considering there are a lot of changes.

File: runtime/Java/pom.xml (hunk 2 above)
Comment (ericvergnaud) [reply runtime/Java/pom.xml:43]: runtime-testsuite is about end-to-end testing i.e. from a grammar and input ensure the output is correct.
it tests the runtime in context

unit-testing a BitSet has nothing to do with end-to-end testing.
unit testing of runtime-specific classes should be with their respective modules.
In antlr4 you'll find that Dart, JS, Python and Swift unit tests are with the runtime.
So this is fixing an inconsistency, not creating one

(I suspect tool-testsuite sits in a dedicated module because it has dependencies on runtime and runtime-testsuite)

**Comment by ericvergnaud** [comment]
All checks now pass.
@KvanTTT I'll wait for #30 and #31 to be merged.

**Comment by KvanTTT** [comment]
> I'm happy to merge https://github.com/antlr/antlr5/pull/30 and https://github.com/antlr/antlr5/pull/31 before this but they don't show as ready ?

Ok, next time I will explicitely press a button for review re-requesting.

File: runtime/Java/pom.xml (hunk 2 above)
Comment (KvanTTT) [reply runtime/Java/pom.xml:43]: > So this is fixing an inconsistency, not creating one

In this case all tests from `tool-testsuite` should be moved to `runtime/Java` tests. But I suggest postponing until the next time.

> I suspect tool-testsuite sits in a dedicated module because it has dependencies on runtime and runtime-testsuite

Actually `tool` has dependency on runtime `runtime/Java`. In the future I suggest extracting common part to a separated module and remove runtime dependency out of `tool` because it's redundant.

File: runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/tree/xpath/XPathTokenAnywhereElement.kt (hunk 1 above)
Comment (KvanTTT) [reply runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/tree/xpath/XPathTokenAnywhereElement.kt:6]: Is it strong need in that for now? What does @kaby76 think about that?

File: runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/atn/ATNSerializer.kt
Comment (KvanTTT) [inline runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/atn/ATNSerializer.kt:1]: Moving to `_unused` is incorrect if want to use `runtime/Kotlin` instead of `runtime/Java` in the `tool` (in the future).

File: runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/tree/xpath/XPathTokenAnywhereElement.kt (hunk 1 above)
Comment (KvanTTT) [reply runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/tree/xpath/XPathTokenAnywhereElement.kt:6]: And `unused` files are not just removed? Actually, I suggest reverting changes related to `unused` package name because maybe they are unused for now (for instance `ATNSerializer`) but they could be useful in the future when we port `tool` to Kotlin. There is no strong need to have such a package name and it's confusing.

File: tool-testsuite/test/org/antlr/v5/test/tool/TestXPath.java [hunk 7]
```diff
@@ -26,6 +28,7 @@
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertNotNull;
 
+@Disabled("Since we plan to no longer support XPath")
```
Comment (KvanTTT) [inline tool-testsuite/test/org/antlr/v5/test/tool/TestXPath.java:30]: Could you please figure out why they are failing? I don't think it's a good idea to disable tests only because they don't work by some reason after some changes. I suspect it relates to a change around using `CharStreams` instead of `ANTLRInputStream`.

File: runtime/Java/src/test/java/org/antlr/v5/runtime/TestCodePointCharStream.java (hunk 3 above)
Comment (ericvergnaud) [reply runtime/Java/src/test/java/org/antlr/v5/runtime/TestCodePointCharStream.java]: Not really a change, just a different way of calling it due to a temporary dependency issue

Comment (ericvergnaud) [reply runtime/Java/src/test/java/org/antlr/v5/runtime/TestCodePointCharStream.java]: Fixed in upcoming commit

File: runtime-testsuite/test/org/antlr/v5/test/runtime/java/api/TestVisitors.java (hunk 6 above)
Comment (ericvergnaud) [reply runtime-testsuite/test/org/antlr/v5/test/runtime/java/api/TestVisitors.java:29]: Yeah I would have if I had thought about it before engaging this split. Then as part of the split I find myself having to decide what to do with this obsolete stuff, and it's obviously better to prune it

**Comment by ericvergnaud** [comment]
Re Kotlin and KMP, since we've opted to use them as the basis for the core runtime, we're all gonna need to become good at it. I expect hiccups (this PR isn't one of them). I believe we should accept them as part of the journey.

**Comment by KvanTTT** [comment]
I haven't had time to re-review this PR yet and you merged it without any approvement.

Sorry, it looks like I not very useful for ANTLR 5 since we have different opinions on code quality and processes.

I'm going to prepare some my old PRs and after that I will work on my own fork. Feel free to develop ANTLR 5 by yourself.

**Comment by ericvergnaud** [comment]
@KvanTTT sorry but I did implement some of the changes you requested. We disagree on priorities, and I'm afraid a PR is not a place to discuss priorities (I don't think #32 and #33 had any priority but I accept different opinions, and I merged them). Despite our different opinions, you are extremely useful to antlr5 and you shouldn't take any of this personally. Feel free to contribute you own way, you're always welcome.

**Comment by KvanTTT** [comment]
> @KvanTTT sorry but I did implement some of the changes you requested.

But I have other notes and I didn't have time to review it again: this PR appeared yesterday. I don't think it's urgent and it had to be merged ASAP. Also, @lppedd didn't accept it.

**Review by KvanTTT — COMMENTED** [review]
I've decided to complete my review anyway despite the PR is already merged.

File: runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/atn/ATNSerializer.kt
Comment (ericvergnaud) [reply runtime/Kotlin/src/org/antlr/v5/runtime/kotlin/_unused/atn/ATNSerializer.kt:1]: Then it will be time to move them to the tool ? Or change how the tool works ? I think we need to progress the serialization specs to decide what to do with this. IMHO we should use an existing xpath engine, not maintain our own (same for template engine)

File: tool-testsuite/test/org/antlr/v5/test/tool/TestXPath.java (hunk 7 above)
Comment (ericvergnaud) [reply tool-testsuite/test/org/antlr/v5/test/tool/TestXPath.java:30]: The idea is to only fix them if we decide to keep our own xpath engine.
